crates/interface-types/src/lib.rs

```diff
@@ -300,42 +344,110 @@ fn abi2ast(param: &ir::AbiParam) -> Result<ast::WebidlScalarType> {
         let alloc_args = vec![RuntimeValue::I32(len)];
-        let results = match cx.invoke(handle, alloc_func_name, &alloc_args)? {
+        let results = match self.cx.invoke(self.handle, alloc_func_name, &alloc_args)? {
             ActionOutcome::Returned { values } => values,
             ActionOutcome::Trapped { message } => bail!("trapped: {}", message),
         };
         if results.len() != 1 {
             bail!("allocator function wrong number of results");
         }
-        let ptr = match results[0] {
+        Ok(match results[0] {
             RuntimeValue::I32(i) => i,
             _ => bail!("allocator function bad return type"),
-        };
-        let memory = handle
+        })
+    }
+    unsafe fn get_memory(&mut self) -> Result<&mut [u8]> {
+        let memory = self
+            .handle
             .lookup("memory")
             .ok_or_else(|| format_err!("no exported `memory`"))?;
         let definition = match memory {
             wasmtime_runtime::Export::Memory { definition, .. } => definition,
             _ => bail!("export `memory` wasn't a `Memory`"),
         };
+        Ok(slice::from_raw_parts_mut(
+            (*definition).base,
+            (*definition).current_length,
+        ))
+    }
+}
+
+struct InstanceTranslateContext(pub api::HostRef<api::Instance>);
+
+impl TranslateContext for InstanceTranslateContext {
+    fn invoke_alloc(&mut self, alloc_func_name: &str, len: i32) -> Result<i32> {
+        let alloc = self
+            .0
+            .borrow()
+            .find_export_by_name(alloc_func_name)
+            .ok_or_else(|| format_err!("failed to find alloc function `{}`", alloc_func_name))?
+            .func()
+            .ok_or_else(|| format_err!("`{}` is not a (alloc) function", alloc_func_name))?
+            .clone();
+        let alloc_args = vec![api::Val::I32(len)];
+        let results = match alloc.borrow().call(&alloc_args) {
+            Ok(values) => values,
+            Err(trap) => bail!("trapped: {:?}", trap),
+        };
+        if results.len() != 1 {
+            bail!("allocator function wrong number of results");
+        }
+        Ok(match results[0] {
+            api::Val::I32(i) => i,
+            _ => bail!("allocator function bad return type"),
+        })
+    }
+    unsafe fn get_memory(&mut self) -> Result<&mut [u8]> {
+        let memory = self
+            .0
+            .borrow()
+            .find_export_by_name("memory")
+            .ok_or_else(|| format_err!("failed to find `memory` export"))?
+            .memory()
+            .ok_or_else(|| format_err!("`memory` is not a memory"))?
+            .clone();
+        let ptr = memory.borrow().data_ptr();
+        let len = memory.borrow().data_size();
+        Ok(std::slice::from_raw_parts_mut(ptr, len))
+    }
+}
+
+fn translate_incoming(
+    cx: &mut dyn TranslateContext,
+    bindings: &[ast::IncomingBindingExpression],
+    args: &[Value],
+) -> Result<Vec<RuntimeValue>> {
```

Comment:
Instead of working with `RuntimeValue` here should this be working with `api::Val`?

Reply:
The `RuntimeValue` will be replaced by `api::Val` once `ModuleData::invoke` will be removed (used by wasmtime-py?)

Reply:
Ah ok, did you want to do the wasmtime-py conversion as part of this PR or save it for a later one?

Reply:
Removing wasmtime_jit / wasmtime_runtime dependecies as well as `RuntimeValue` depends on wasmtime-py. (Is it the only interface types user left?) I did not plan to do it in this PR, but depending on timing and feedback, I can change wasmtime-py and remove low-level stuff here.